src/Arara/Process/IPC/Semaphore.php

```diff
@@ -0,0 +1,132 @@
+<?php
+
+namespace Arara\Process\IPC ;
+
+/**
+ * This class provides Interprocess communication using UNIX Semaphores
+ *
+ * Code written by Nikolay and copied into Arara/Process under the terms of
+ * the MIT license.
+ *
+ * Full credit to Nikolay for his excellent class
+ * 
+ * @license http://opensource.org/licenses/MIT
+ * @author Nikolay Bondarenko http://misterion.ru 
+ */
+class Semaphore
+{
+    /**
+     * @var resource
+     */
+    protected $mutex;
+
+    /**
+     * @var string
+     */
+    protected $file;
+
+    /**
+     * @var bool
+     */
+    protected $isAcquired;
+
+    /**
+     * Creator process pid
+     *
+     * @var int
+     */
+    protected $pid;
+
+    public function __construct($key = null)
+    {
+        $this->pid = getmypid();
+
+        $this->isAcquired = false;
+        if ($key) {
+            $semKey = $key;
+        } else {
+            $this->file = tempnam(sys_get_temp_dir(), 's');
+            $semKey = ftok($this->file, 'a');
+        }
+
+        $this->mutex = sem_get($semKey, 1); //auto_release = 1 by default
+        if (!$this->mutex) {
+            throw new \RuntimeException('Unable to create the semaphore');
+        }
+    }
+
+    public function __destruct()
+    {
+        if ($this->isAcquired()) {
+            $this->release();
+        }
+
+        if ($this->pid === getmypid()) {
+            $this->remove();
+        }
```

Comment:
Don't worry, there is no problem at all and there is no need to re-open the PR either. :smile: 

IPC is not required to this project, since using signals is enough but this is feature is a _nice-to-have_, of course, that's why I created #2.

In a near future I hope I will work on that.